runtime/gc_base/ObjectAccessBarrier.cpp

```diff
@@ -1410,6 +1410,68 @@ MM_ObjectAccessBarrier::cloneObject(J9VMThread *vmThread, J9Object *srcObject, J
 	return;
 }
 
+#if defined(J9VM_OPT_VALHALLA_VALUE_TYPES)
+/**
+ * Copy all of the fields of a value class instance to another value class instance.
+ * The source or destination may be a flattened value within another object, meaning
+ * srcOffset and destOffset need not be equal. This is based on cloneObject(...).
+ * @TODO This does not currently check if the fields that it is reading are volatile.
+ *
+ * @oaram valueClass The value class.
+ * @param srcObject The object containing the value class instance fields being copied.
+ * @param srcOffset The offset of the value class instance fields in srcObject.
+ * @param destValue The object containing the value class instance fields being copied to.
+ * @param destOffset The offset of the value class instance fields in destObject.
+ */
+void
+MM_ObjectAccessBarrier::copyValue(J9VMThread *vmThread, J9Class *valueClass, J9Object *srcObject, UDATA srcOffset, J9Object *destObject, UDATA destOffset)
+{
+	/* We currently do not make a distinction between values that only contain
+	 * primitives and values that may contain a reference (ie. value vs mixed-value
+	 * in packedObject terminology). As a result, we will treat all values as if
+	 * they may contain references. In the future this may change.
+	 *
+	 * Value types have no need or lockwords, however, they are still present in the
+	 * current implementation. For now we will just skip over them by specifying
+	 * appropriate offsets. We will also skip over the bit in the instance description.
+	 */
+	const UDATA *descriptionPtr = (UDATA *) valueClass->instanceDescription;
+	UDATA descriptionBits;
+	if(((UDATA)descriptionPtr) & 1) {
+		descriptionBits = ((UDATA)descriptionPtr) >> 1;
+	} else {
+		descriptionBits = *descriptionPtr++;
+	}
+
+	UDATA descriptionIndex = J9_OBJECT_DESCRIPTION_SIZE - 1;
+	UDATA offset = 0;
+	UDATA limit = valueClass->totalInstanceSize;
+
+	if (0 <= (IDATA)valueClass->lockOffset) {
+		/* skip lockword */
+		limit -= sizeof(j9objectmonitor_t);
+		descriptionBits >>= 1;
+		descriptionIndex -= 1;
+	}
+
+	while (offset < limit) {
+		/* Determine if the slot contains an object pointer or not */
+		if(descriptionBits & 1) {
+			J9Object *objectPtr = mixedObjectReadObject(vmThread, srcObject, srcOffset + offset, false);
+			mixedObjectStoreObject(vmThread, destObject, destOffset + offset, objectPtr, false);
+		} else {
+			*(fj9object_t *)((UDATA)destObject + destOffset + offset) = *(fj9object_t *)((UDATA)srcObject + srcOffset + offset);
+		}
+		descriptionBits >>= 1;
+		if(descriptionIndex-- == 0) {
+			descriptionBits = *descriptionPtr++;
+			descriptionIndex = J9_OBJECT_DESCRIPTION_SIZE - 1;
+		}
+		offset += sizeof(fj9object_t);
+	}
+}
+#endif /* defined(J9VM_OPT_VALHALLA_VALUE_TYPES) */
+
```

Comment:
Am I right in observing this is a subset of that cloneObject does. Could we share this code within a private method and call it both by cloneObject (with src/dest offsets being 0) and copyValue?